parsec/mca/device/cuda/device_cuda_module.c

```diff
@@ -1131,9 +1137,15 @@ parsec_gpu_data_reserve_device_space( parsec_device_cuda_module_t* cuda_device,
 #else
         gpu_elem = lru_gpu_elem;
 #endif
-        assert( 0 == gpu_elem->readers );
+
+        /* Do not push it back into the LRU for now to prevent others from discovering
+         * this copy and trying to acquire it. If we fail to find all the copies we need
+         * we will push it back in the release_temp_and_return, otherwise they will become
+         * available once properly updated.
+         */
         gpu_elem->coherency_state = PARSEC_DATA_COHERENCY_INVALID;
-        gpu_elem->version = 0;
+        gpu_elem->version = UINT_MAX;  /* scrap value for now */
+        gpu_elem->readers = 0;         /* safe to set to no access */
```

Comment:
I'm not sure this is safe: there might be someone looking at the copy seeing negative readers and decrementing by one. I think this should be an atomic increment by `INT_MAX` so that eventually `readers` reaches 0 in all cases.

Reply:
Or an atomic increment by `-INT_MIN` ? (`INT_MIN` is probably not equal to `INT_MAX`)

Reply:
At this point all the protection that could be offered was offered by the atomic update protocols. Now, I know this code remains unsafe for extreme cases, but the only way to fix it is to add a lock. Adding yet another atomic operation will not help.

If another thread is trying to use this copy as a candidate for d2d, and that thread is delayed, it might see the readers set to 0 and assume everything is fine and then move ahead and use the copy. Hopefully that will not happen because the version will be incorrect and the coherency is invalid.

Reply:
This assignment conflicts with the roll back a potential reader would do here: https://github.com/ICLDisco/parsec/pull/575/files#diff-a42087f8b140e7fc55fd74480d49a4334d21ff2810f51ca57e0d1955534286e8R1415

The sequence would be: Thread A sets `readers` to `INT_MIN`; Thread B increments `readers` and finds it negative; Thread A detaches the copy and sets `readers` to 0; Thread B does the rollback by atomically decrementing `readers`. The result would be `readers == -1`. I don't think that should be valid. 

Instead of assigning 0 we should atomically add `-INT_MIN` which would guarantee `readers >= 0`.